### PR TITLE
Bug fixes and test improvements

### PR DESCRIPTION
**Pull request by apaszke**
* Add same device asserts for cuDNN (#720)
* Fix bug in `Module.zero_grad` (#719)
* Add CUDA double and half THNN backends (#716, #48)
* Add output shape checks in conv (#704)
* cuDNN RNN can be now backproped through multiple times (#703)
* Fix auto-gpu in `torch.cat` (#689)
* Add tests for non-contiguous inputs and gradients in nn
* Fixed some flaky tests
* Added some more asserts to cuDNN RNNs

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -380,7 +380,8 @@ def eval(self):
     def zero_grad(self):
         """Sets gradients of all model parameters to zero."""
         for p in self.parameters():
-            p.grad.data.zero_()
+            if p.grad is not None:
```
Comment (colesbury) [inline torch/nn/modules/module.py]: I think this should be `p._grad is not None` (otherwise it's always True)

File: torch/nn/_functions/conv.py [hunk 2]
```diff
@@ -27,6 +27,8 @@ def __init__(self, stride, padding, dilation, transposed, output_padding,
         self.groups = groups
 
     def forward(self, input, weight, bias=None):
+        if not input.is_contiguous():
+            raise ValueError("Convolution input has to be contiguous")
```
Comment (soumith) [inline torch/nn/_functions/conv.py]: until the cudnn bindings are changed to use separate descriptors for input and gradOutput, this is presumably needed.
